### PR TITLE
[WIP][OpenMP] Codegen & Driver changes for AMDGPU offloading

### PR DESCRIPTION
**Pull request by pdhaliwal-amd**
I was able to get smoke/helloworld work. Running other smoke tests as well.

### CODE REVIEW

File: clang/include/clang/Driver/Driver.h [hunk 1]
```diff
@@ -254,6 +254,9 @@ class Driver {
   /// stored in it, and will clean them up when torn down.
   mutable llvm::StringMap<std::unique_ptr<ToolChain>> ToolChains;
 
+  /// Number of parallel jobs.
+  unsigned NumParallelJobs;
+
```
Comment (pdhaliwal-amd) [inline clang/include/clang/Driver/Driver.h]: This might not be required for now.

File: clang/lib/CodeGen/CGCall.cpp [hunk 2]
```diff
@@ -742,8 +742,9 @@ CodeGenTypes::arrangeLLVMFunctionInfo(CanQualType resultType,
                                       FunctionType::ExtInfo info,
                      ArrayRef<FunctionProtoType::ExtParameterInfo> paramInfos,
                                       RequiredArgs required) {
-  assert(llvm::all_of(argTypes,
-                      [](CanQualType T) { return T.isCanonicalAsParam(); }));
+  if (!getContext().getLangOpts().OpenMP)
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGCall.cpp]: Without this map does not work

File: clang/lib/CodeGen/CGOpenMPRuntime.h [hunk 3]
```diff
@@ -495,6 +495,9 @@ class CGOpenMPRuntime {
   ///   int32_t flags;
   ///   int32_t reserved;
   /// };
+  //
+  QualType TgtAttributeStructQTy;
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntime.h]: This can be removed after MaxParallelLevel is removed.

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp [hunk 4]
```diff
@@ -1099,13 +1197,23 @@ void CGOpenMPRuntimeGPU::emitNonSPMDKernel(const OMPExecutableDirective &D,
...
-        llvm::GlobalValue::InternalLinkage,
-        llvm::ConstantPointerNull::get(CGM.VoidPtrTy),
-        "_openmp_kernel_static_glob_rd$ptr", /*InsertBefore=*/nullptr,
-        llvm::GlobalValue::NotThreadLocal,
-        CGM.getContext().getTargetAddressSpace(LangAS::cuda_shared));
+    KernelStaticGlobalized =
+        (CGM.getTriple().isAMDGCN())
+            ? new llvm::GlobalVariable(
+                  CGM.getModule(), CGM.VoidPtrTy, /*isConstant=*/false,
+                  llvm::GlobalValue::WeakAnyLinkage,
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Why is InternalLinkage not working?

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp [hunk 5]
```diff
@@ -1231,17 +1341,29 @@ void CGOpenMPRuntimeGPU::emitSPMDKernel(const OMPExecutableDirective &D,
...
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Same as above

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp [hunk 6]
```diff
@@ -2105,6 +2292,8 @@ void CGOpenMPRuntimeGPU::emitNonSPMDParallelCall(
   Address ZeroAddr = CGF.CreateDefaultAlignTempAlloca(CGF.Int32Ty,
                                                       /*Name=*/".zero.addr");
   CGF.InitTempAlloca(ZeroAddr, CGF.Builder.getInt32(/*C*/ 0));
+
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Extra

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp [hunk 7]
```diff
@@ -2140,19 +2333,35 @@ void CGOpenMPRuntimeGPU::emitNonSPMDParallelCall(
     RCG(CGF);
   };
 
-  auto &&L0ParallelGen = [this, CapturedVars, Fn](CodeGenFunction &CGF,
-                                                  PrePostActionTy &Action) {
+  bool IsAmdgcn = (CGM.getTriple().isAMDGCN()) ? true : false;
+
+  auto &&L0ParallelGen = [this, CapturedVars, Fn, IsAmdgcn](
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: IsAmdgcn capture is redundant.

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp [hunk 8]
```diff
@@ -2280,6 +2495,8 @@ void CGOpenMPRuntimeGPU::emitSPMDParallelCall(
   Address ZeroAddr = CGF.CreateDefaultAlignTempAlloca(CGF.Int32Ty,
                                                       /*Name=*/".zero.addr");
   CGF.InitTempAlloca(ZeroAddr, CGF.Builder.getInt32(/*C*/ 0));
+
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Remove this

File: clang/lib/CodeGen/CodeGenAction.cpp [hunk 9]
```diff
@@ -980,20 +981,67 @@ CodeGenAction::CreateASTConsumer(CompilerInstance &CI, StringRef InFile) {
         return nullptr;
       }
 
-      Expected<std::unique_ptr<llvm::Module>> ModuleOrErr =
-          getOwningLazyBitcodeModule(std::move(*BCBuf), *VMContext);
-      if (!ModuleOrErr) {
-        handleAllErrors(ModuleOrErr.takeError(), [&](ErrorInfoBase &EIB) {
+      if (StringRef(F.Filename).endswith(".a")) {
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CodeGenAction.cpp]: Will test if this is required (for now).

**Review by pdhaliwal-amd — COMMENTED** [review]
I have posted some questions here for which I need answers. I will keep on removing the stuff not required.

File: clang/include/clang/Driver/Driver.h (hunk 1 above)
Comment (JonChesterfield) [reply clang/include/clang/Driver/Driver.h]: Agreed, don't think it's necessary for functionality

File: clang/lib/CodeGen/CGCall.cpp (hunk 2 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/CGCall.cpp]: suspicious, wonder why we differ from nvptx here. Might be address space qualification related

File: clang/lib/CodeGen/CGOpenMPRuntimeNVPTX.cpp [hunk 10]
```diff
@@ -33,6 +33,13 @@ CGOpenMPRuntimeNVPTX::CGOpenMPRuntimeNVPTX(CodeGenModule &CGM)
 }
 
 llvm::Value *CGOpenMPRuntimeNVPTX::getGPUWarpSize(CodeGenFunction &CGF) {
+  if (CGF.getTarget().getTriple().isAMDGCN()) {
```
Comment (JonChesterfield) [inline clang/lib/CodeGen/CGOpenMPRuntimeNVPTX.cpp]: This is probably a missing patch in trunk - getGPUWarpSize should resolve to nvptx or amdgcn in the same fashion as the other factored out functions

File: clang/lib/CodeGen/CodeGenAction.cpp (hunk 9 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/CodeGenAction.cpp]: Nope, none of our stuff requires linking archive formats. I think libm used to. Can be split out from this

File: clang/lib/CodeGen/TargetInfo.cpp [hunk 11]
```diff
@@ -9015,14 +9015,26 @@ void AMDGPUTargetCodeGenInfo::setTargetAttributes(
...
+  // Move the GV->isDeclaration() check till after properly setting
+  // device function attributes.
+
   const FunctionDecl *FD = dyn_cast_or_null<FunctionDecl>(D);
   if (!FD)
     return;
 
   llvm::Function *F = cast<llvm::Function>(GV);
 
+  if (M.getLangOpts().OpenMP) {
```
Comment (JonChesterfield) [inline clang/lib/CodeGen/TargetInfo.cpp:9014]: This sort of toggling O0/noinline/alwaysinline might be related to the function pointer hashing, in which case we can drop it in favour of fixed-function-abi for trunk

File: clang/lib/Driver/Driver.cpp [hunk 12]
```diff
@@ -2465,8 +2484,9 @@ class OffloadingActionBuilder final {
           return ABRT_Success;
 
         // Replicate inputs for each GPU architecture.
-        auto Ty = IA->getType() == types::TY_HIP ? types::TY_HIP_DEVICE
-                                                 : types::TY_CUDA_DEVICE;
+        auto Ty = AssociatedOffloadKind == Action::OFK_HIP
```
Comment (JonChesterfield) [inline clang/lib/Driver/Driver.cpp]: some of this looks like a spurious delta between trunk and amd-stg-open, perhaps we should propose a (mostly whitespace) patch to amd-stg-open based on trunk to get the signal/noise up

File: clang/lib/Driver/Driver.cpp [hunk 13]
```diff
@@ -2945,6 +2966,12 @@ class OffloadingActionBuilder final {
     /// The OpenMP actions for the current input.
     ActionList OpenMPDeviceActions;
 
+    bool CompileHostOnly = false;
```
Comment (JonChesterfield) [inline clang/lib/Driver/Driver.cpp:2979]: This device-only openmp compilation is likely to be needed to build the deviceRTL from openmp, instead of from cuda + hip.

File: clang/lib/Driver/Driver.cpp [hunk 14]
```diff
@@ -2964,6 +2991,26 @@ class OffloadingActionBuilder final {
       assert(OpenMPDeviceActions.size() == ToolChains.size() &&
              "Number of OpenMP actions and toolchains do not match.");
 
+      // FIXME:  Is there an easier way to determine amdgcn at this point?
```
Comment (JonChesterfield) [inline clang/lib/Driver/Driver.cpp:3004]: think we have access to the offload triple here, should check that

File: clang/lib/Driver/Driver.cpp [hunk 15]
```diff
@@ -3036,7 +3105,12 @@ class OffloadingActionBuilder final {
         for (Action *&A : OpenMPDeviceActions) {
           assert(isa<CompileJobAction>(A));
           OffloadAction::DeviceDependences DDep;
-          DDep.add(*A, **TC, /*BoundArch=*/nullptr, Action::OFK_OpenMP);
+          if (GpuArchList.size())
```
Comment (JonChesterfield) [inline clang/lib/Driver/Driver.cpp:3099]: i think this is about compiling for multiple gpu arch at the same time. if trunk doesn't do that for nvptx, we don't need to do it for amdgcn as part of the minimum viable patch

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 16]
```diff
@@ -554,6 +558,44 @@ void tools::addLTOOptions(const ToolChain &ToolChain, const ArgList &Args,
   addX86AlignBranchArgs(D, Args, CmdArgs, /*IsLTO=*/true);
 }
 
+std::string tools::FindDebugInLibraryPath() {
```
Comment (JonChesterfield) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp]: the find-debug-libraries feature is probably aomp specific, at least for now

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 17]
```diff
@@ -1068,7 +1114,7 @@ tools::ParsePICArgs(const ToolChain &ToolChain, const ArgList &Args) {
   }
 
   // AMDGPU-specific defaults for PIC.
-  if (Triple.getArch() == llvm::Triple::amdgcn)
+  if (Triple.isAMDGCN())
```
Comment (JonChesterfield) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp]: another one for the noise reduction patch against amd-stg-open

File: clang/lib/Driver/ToolChains/Cuda.cpp [hunk 18]
```diff
@@ -509,6 +509,9 @@ void NVPTX::OpenMPLinker::ConstructJob(Compilation &C, const JobAction &JA,
   assert(TC.getTriple().isNVPTX() && "Wrong platform");
 
   ArgStringList CmdArgs;
+  const char *Exec =
```
Comment (JonChesterfield) [inline clang/lib/Driver/ToolChains/Cuda.cpp]: this stuff is a fix for nvptx

**Review by JonChesterfield — COMMENTED** [review]
Interesting diff, thanks. I've left inline comments on the bits I recognise. Optimistic that this can get much smaller

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp (hunk 4 above)
Comment (JonChesterfield) [reply clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Unclear. Works fine on gfx906, failed on gfx803 in difficult to debug fashion.

File: clang/lib/CodeGen/CGOpenMPRuntime.h (hunk 3 above)
Comment (pdhaliwal-amd) [reply clang/lib/CodeGen/CGOpenMPRuntime.h]: Cannot be removed

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp [hunk 19]
```diff
@@ -2324,6 +2541,12 @@ void CGOpenMPRuntimeGPU::emitSPMDParallelCall(
     // If we are not in the target region, it is definitely L2 parallelism or
     // more, because for SPMD mode we always has L1 parallel level, sowe don't
     // need to check for orphaned directives.
+    if (CGM.getTriple().isAMDGCN()) {
+      unsigned DiagID = CGM.getDiags().getCustomDiagID(
+        DiagnosticsEngine::Remark,
+        "Nested parallel pragma, this will be serialized on device");
+      CGM.getDiags().Report(DiagID);
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Is this required?

File: clang/lib/Driver/ToolChains/CommonArgs.cpp (hunk 17 above)
Comment (pdhaliwal-amd) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp]: Removed

File: clang/lib/Driver/ToolChains/Cuda.cpp (hunk 18 above)
Comment (pdhaliwal-amd) [reply clang/lib/Driver/ToolChains/Cuda.cpp]: Removed

File: clang/lib/Driver/ToolChains/HIP.cpp [hunk 20]
```diff
@@ -88,6 +89,14 @@ void AMDGCN::Linker::constructLldCommand(Compilation &C, const JobAction &JA,
         Args.MakeArgString(Twine("-plugin-opt=") + A->getValue(0)));
   }
 
+  Arg *GTune =
```
Comment (pdhaliwal-amd) [inline clang/lib/Driver/ToolChains/HIP.cpp]: Seems HIP specific change. Removing

File: clang/lib/CodeGen/TargetInfo.cpp (hunk 11 above)
Comment (pdhaliwal-amd) [reply clang/lib/CodeGen/TargetInfo.cpp:9014]: What is fixed-function-abi?

**Review by pdhaliwal-amd — COMMENTED** [review]
Removed more stuff. Mostly HIP and CUDA related changes. There are some things I am trying to check why those were required originally. For e.g. assert one.

**Comment by ronlieb** [comment]
Validating these sorts of changes (ie code removal) is a bit tedious, in the past i would apply them to a local build of amd-stg-open, do a full build, and then run ALL the tests looking for corner case issues. including spec accel (offload), and spec hpg(offload).

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp (hunk 4 above)
Comment (pdhaliwal-amd) [reply clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Thinking of removing this unless you think otherwise. On gfx908, program compiles and run successfully with older version

File: clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp [hunk 21]
```diff
@@ -586,6 +588,13 @@ static llvm::Value *getMasterThreadID(CodeGenFunction &CGF) {
                        Bld.CreateNot(Mask), "master_tid");
 }
 
+static void DotToUnderbar(llvm::Function *Fn) {
```
Comment (pdhaliwal-amd) [inline clang/lib/CodeGen/CGOpenMPRuntimeGPU.cpp]: Remove this
